### DISCUSSION TITLE
Metal 3?

### DISCUSSION THREAD

**Discussion by ajmas** [LWJGL/lwjgl3#811] [General]
Following the announcement of [Metal 3](https://developer.apple.com/metal/), will this have much impact on the bindings on macOS? Are there any plans around this?

BTW I became aware of Metal 3, because of a [feature request ](https://feedback.minecraft.net/hc/en-us/community/posts/6780237447821-Metal-3-API-for-Minecraft-Java-)for Minecraft.

**Comment by SWinxy** [comment]
Metal can be accessed via LWJGL3's [macOS stuff](https://javadoc.lwjgl.org/org/lwjgl/system/macosx/package-summary.html), though it doesn't serve as bindings to Metal since it's a "figure it out yourself" interface to the Objective-C runtime. Mojang is far better off switching to Vulkan for its graphics API than adding a Metal option, since Vulkan is supported natively via [a Metal bridge](https://github.com/KhronosGroup/MoltenVK) that LWJGL provides out-of-the-box.
No plans for adding Metal, though it's not like that would be difficult to wrap in pure Java.

If you want to do this yourself, you would start with something like the code snippet below:
```java
SharedLibrary objc = ObjCRuntime.getLibrary(); // retrieves the 'shared library' object
long msgSend = objc.getFunctionAddress("objc_msgSend"); // gets the function pointer to the (ObjC) runtime's msgSend function
SharedLibrary mtk = MacOSXLibrary.create("/System/Library/Frameworks/MetalKit.framework");
SharedLibrary lib = MacOSXLibrary.create("/System/Library/Frameworks/Metal.framework");
```

Those should be everything one would need to work with Metal using Java. Though again, you'd need to know Objective-C, the Objective-C Runtime, and Metal. Not *im*possible, but not everything is evident. lwjgl3-awt has some [example code](https://github.com/LWJGLX/lwjgl3-awt/blob/main/src/org/lwjgl/vulkan/awt/PlatformMacOSXVKCanvas.java) if you need it for reference, which includes annotations for what its doing and its Objective-C analogue.
